{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys \n",
    "sys.path.append(\"../../\")\n",
    "from QUEEN.queen import *\n",
    "if \"output\" not in os.listdir(\"./\"):\n",
    "    os.mkdir(\"output\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "#### Example code 1: : Create a QUEEN object for a blunt-ended dsDNA \n",
    "The top-stranded sequence (5'-to-3') of the generating QUEEN object is given. In default, the sequence ends and topology will be interpreted as blunt-ends and linear, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = QUEEN(seq=\"CCGGTATGCGTCGA\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "#### Example code 2: Create a QUEEN object for a sticky-ended dsDNA\n",
    "The left and right values separated by \"/\" show the top and bottom strand sequences of the generating QUEEN object, respectively. The top strand sequence is given in 5’-to-3’ direction from left to right and the bottom strand sequence is given by 3’-to-5’ direction from left to right. \"-\" indicates gap. A:T and G:C base-pairing rule is required between the two strings except for the gap positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = QUEEN(seq=\"CCGGTATGCG----/----ATACGCAGCT\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "#### Example code 3: Create a QUEEN object for a circular dsDNA\n",
    "The sequence topology of generating QUEEN object can be given by \"linear\" or \"circular\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dna = QUEEN(seq=\"CCGGTATGCGTCGA\", topology=\"circular\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "----\n",
    "#### Example code 4.1: Create an annotated DNA object from GenBank format file  \n",
    "The local file path for GenBank input of generating QUEEN object is given."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "plasmid  = QUEEN(record=\"input/pX330.gbk\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example code 4.2: Create a QUEEN object from a NCBI Accession ID\n",
    "A NCBI accession number for the GenBank input of generating QUEEN object is given. When entering NCBI accession number, dbtype parameter should be given by “ncbi” as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download pUC19 plasmid GenBank input from ncbi \n",
    "pUC19 = QUEEN(record=\"M77789.2\", dbtype=\"ncbi\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example code 4.3: Create a QUEEN object from an Addgene plasmid ID\n",
    "An Addgene plasmid ID for the GenBank input of generating QUEEN object is given. When entering Addgene plasmid ID, dbtype should be given by “addgene” as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Download pUC19 plasmid GenBank input from ncbi \n",
    "pUC19 = QUEEN(record=\"50005\", dbtype=\"addgene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example code 4.4: Create a QUEEN object from a Benchling share link\n",
    "A Benchling share link for the GenBank input of generating QUEEN object is given. When entering Benchling share links, dbtype should be given by “benchling” as follows. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "plasmid = QUEEN(record=\"https://benchling.com/s/seq-U4pePb09KHutQzjyOPQV\", dbtype=\"benchling\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above URL linked to the sequence file of pX330 plasmid located in my Benchling directory is given. The pX330 plasmid containing a Cas9 gene and a gRNA expression cassette is a cloning vector used for genome editing experiments.   \n",
    "In the following example codes, the plasmid variable represents the QUEEN object of the pX330 plasmid."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "#### Example code 5: Print a dsDNA sequence of sticky ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5' CCGGTATGCG---- 3'\n",
      "3' ----ATACGCAGCT 5'\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CCGGTATGCG----/----ATACGCAGCT'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna = QUEEN(seq=\"CCGGTATGCG----/----ATACGCAGCT\") \n",
    "dna.printsequence(display=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "#### Example code 6: Print DNA features with a well-formatted table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_id  feature_type   qualifier:label     start  end   strand  \n",
      "1           source         source              0      8484  +       \n",
      "100         primer_bind    hU6-F               0      21    +       \n",
      "200         promoter       U6 promoter         0      241   +       \n",
      "300         primer_bind    LKO.1 5'            171    191   +       \n",
      "400         misc_RNA       gRNA scaffold       267    343   +       \n",
      "500         enhancer       CMV enhancer        439    725   +       \n",
      "600         intron         hybrid intron       983    1211  +       \n",
      "700         regulatory     Kozak sequence      1222   1232  +       \n",
      "800         CDS            3xFLAG              1231   1297  +       \n",
      "900         CDS            SV40 NLS            1303   1324  +       \n",
      "1000        CDS            Cas9                1348   5449  +       \n",
      "1100        CDS            nucleoplasmin NLS   5449   5497  +       \n",
      "1200        primer_bind    BGH-rev             5524   5542  -       \n",
      "1300        polyA_signal   bGH poly(A) signal  5530   5738  +       \n",
      "1400        repeat_region  AAV2 ITR            5746   5876  +       \n",
      "1500        repeat_region  AAV2 ITR            5746   5887  +       \n",
      "1600        rep_origin     f1 ori              5961   6417  +       \n",
      "1700        primer_bind    F1ori-R             6048   6068  -       \n",
      "1800        primer_bind    F1ori-F             6258   6280  +       \n",
      "1900        primer_bind    pRS-marker          6433   6453  -       \n",
      "2000        primer_bind    pGEX 3'             6552   6575  +       \n",
      "2100        primer_bind    pBRforEco           6612   6631  -       \n",
      "2200        promoter       AmpR promoter       6698   6803  +       \n",
      "2300        CDS            AmpR                6803   7664  +       \n",
      "2400        primer_bind    Amp-R               7021   7041  -       \n",
      "2500        rep_origin     ori                 7834   8423  +       \n",
      "2600        primer_bind    pBR322ori-F         8323   8343  +       \n",
      "\n"
     ]
    }
   ],
   "source": [
    "plasmid.printfeature()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "#### Example code 7: Search DNA sequences with a regular expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start  end   strand  sequence                 \n",
      "115    138   +       GTAGAAAGTAATAATTTCTTGGG  \n",
      "523    546   +       GACTTTCCATTGACGTCAATGGG  \n",
      "816    839   +       GTGCAGCGATGGGGGCGGGGGGG  \n",
      "1372   1395  +       GACATCGGCACCAACTCTGTGGG  \n",
      "1818   1841  +       GGCCCACATGATCAAGTTCCGGG  \n",
      "3097   3120  +       GATCGGTTCAACGCCTCCCTGGG  \n",
      "3300   3323  +       GCGGCGGAGATACACCGGCTGGG  \n",
      "3336   3359  +       GAAGCTGATCAACGGCATCCGGG  \n",
      "3529   3552  +       GGCAGCCCCGCCATTAAGAAGGG  \n",
      "3577   3600  +       GACGAGCTCGTGAAAGTGATGGG  \n",
      "3640   3663  +       GAGAACCAGACCACCCAGAAGGG  \n",
      "3697   3720  +       GAAGAGGGCATCAAAGAGCTGGG  \n",
      "3783   3806  +       GTACTACCTGCAGAATGGGCGGG  \n",
      "3915   3938  +       GACCAGAAGCGACAAGAACCGGG  \n",
      "4303   4326  +       GCCTACCTGAACGCCGTCGTGGG  \n",
      "4552   4575  +       GGGGAGATCGTGTGGGATAAGGG  \n",
      "4701   4724  +       GATCGCCAGAAAGAAGGACTGGG  \n",
      "4777   4800  +       GTGGTGGCCAAAGTGGAAAAGGG  \n",
      "5217   5240  +       GTCCGCCTACAACAAGCACCGGG  \n",
      "5653   5676  +       GTAGGTGTCATTCTATTCTGGGG  \n",
      "5679   5702  +       GGGGTGGGGCAGGACAGCAAGGG  \n",
      "5866   5889  +       GAGCGCGCAGCTGCCTGCAGGGG  \n",
      "6158   6181  +       GACCCCAAAAAACTTGATTTGGG  \n",
      "6303   6326  +       GCTATTCTTTTGATTTATAAGGG  \n",
      "6534   6557  +       GACAAGCTGTGACCGTCTCCGGG  \n",
      "6667   6690  +       GACGTCAGGTGGCACTTTTCGGG  \n",
      "6908   6931  +       GATCAGTTGGGTGCACGAGTGGG  \n",
      "7524   7547  +       GCGGTATCATTGCAGCACTGGGG  \n",
      "8082   8105  +       GCGATAAGTCGTGTCTTACCGGG  \n",
      "8135   8158  +       GCGCAGCGGTCGGGCTGAACGGG  \n",
      "8302   8325  +       GCGCACGAGGGAGCTTCCAGGGG  \n",
      "8334   8357  +       GGTATCTTTATAGTCCTGTCGGG  \n",
      "8382   8405  +       GATTTTTGTGATGCTCGTCAGGG  \n",
      "8466   5     +       GGCCTTTTGCTCACATGTGAGGG  \n",
      "493    516   -       GCGTTACTATTGACGTCAATGGG  \n",
      "654    677   -       GTCCCATAAGGTCATGTACTGGG  \n",
      "758    781   -       GGTGGGGAGGGGGGGGAGATGGG  \n",
      "1014   1037  -       GCGCGAGGCGGCGGCGGAGCGGG  \n",
      "1301   1324  -       GACCTTCCGCTTCTTCTTTGGGG  \n",
      "1820   1843  -       GCCCCGGAACTTGATCATGTGGG  \n",
      "2090   2113  -       GAAGTTGCTCTTGAAGTTGGGGG  \n",
      "2183   2206  -       GGCGTACTGGTCGCCGATCTGGG  \n",
      "2288   2311  -       GATCATAGAGGCGCTCAGGGGGG  \n",
      "2689   2712  -       GCCAGAGGGCCCACGTAGTAGGG  \n",
      "3534   3557  -       GGATGCCCTTCTTAATGGCGGGG  \n",
      "3653   3676  -       GCTGTTCTTCTGTCCCTTCTGGG  \n",
      "3741   3764  -       GCAGCTGGGTGTTTTCCACGGGG  \n",
      "4154   4177  -       GTCGTACTTAGTGTTCATCCGGG  \n",
      "4214   4237  -       GGACACCAGCTTGGACTTCAGGG  \n",
      "4295   4318  -       GGCGTTCAGGTAGGCGTCGTGGG  \n",
      "4752   4775  -       GCACAGAATAGGCCACGGTGGGG  \n",
      "4949   4972  -       GCCGTTTTCCAGCTCGAACAGGG  \n",
      "5024   5047  -       GTTCACATATTTGGAGGGCAGGG  \n",
      "5088   5111  -       GTTTCTGCTCATTATCCTCGGGG  \n",
      "5381   5404  -       GGTGATGCTCTGGTGGATCAGGG  \n",
      "5565   5588  -       GTCAAGGAAGGCACGGGGGAGGG  \n",
      "5751   5774  -       GTGGCCAACTCCATCACTAGGGG  \n",
      "5966   5989  -       GCGCTTAATGCGCCGCTACAGGG  \n",
      "6044   6067  -       GGGAAGAAAGCGAAAGGAGCGGG  \n",
      "6206   6229  -       GGGCGAAAAACCGTCTATCAGGG  \n",
      "6467   6490  -       GCGGGTGTTGGCGGGTGTCGGGG  \n",
      "6497   6520  -       GGGAGCAGACAAGCCCGTCAGGG  \n",
      "7771   7794  -       GGAACGAAAACTCACGTTAAGGG  \n",
      "7808   7831  -       GATCCTTTGATCTTTTCTACGGG  \n",
      "\n"
     ]
    }
   ],
   "source": [
    "match_list = plasmid.searchsequence(query=\"G[ATGC]{19}GGG\")\n",
    "plasmid.printfeature(match_list, seq=True, attribute=[\"start\", \"end\", \"strand\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----\n",
    "##### Example code 8: Search DNA sequences with a fuzzy matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_id  feature_type  qualifier:label  start  end   strand  sequence  \n",
      "null        misc_feature  null             5484   5492  +       AAAAAAGA  \n",
      "null        misc_feature  null             6369   6377  +       AACAAAAA  \n",
      "null        misc_feature  null             7872   7880  +       AAACAAAA  \n",
      "null        misc_feature  null             346    354   -       AAAACAAA  \n",
      "null        misc_feature  null             799    807   -       AAAAAATA  \n",
      "null        misc_feature  null             1201   1209  -       GAAAAAAA  \n",
      "null        misc_feature  null             6716   6724  -       AAAAATAA  \n",
      "null        misc_feature  null             7844   7852  -       AGAAAAAA  \n",
      "\n"
     ]
    }
   ],
   "source": [
    "match_list = plasmid.searchsequence(query=\"(?:AAAAAAAA){s<=1}\")\n",
    "plasmid.printfeature(match_list, seq=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---- \n",
    "##### Example code 9: Search DNA sequences with IUPAC degenerate nucleotide codes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_id  feature_type  qualifier:label  start  end   strand  sequence          \n",
      "null        misc_feature  null             4098   4114  +       GAGACAGCTGGTGGAA  \n",
      "null        misc_feature  null             3550   3566  -       CTGTCTGCAGGATGCC  \n",
      "null        misc_feature  null             5239   5255  -       CTCTGATGGGCTTATC  \n",
      "null        misc_feature  null             6415   6431  -       GAGAGTGCACCATAAA  \n",
      "null        misc_feature  null             8357   8373  -       GTCAGAGGTGGCGAAA  \n",
      "\n"
     ]
    }
   ],
   "source": [
    "match_list = plasmid.searchsequence(query=\"SWSWSWDSDSBHBRHH\")\n",
    "plasmid.printfeature(match_list, seq=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "##### Example code 10: Search for specified seqeunce features  \n",
    "Search for DNAfeature objects with a feature type of \"primer_bind\" and retrieve only primers holding the given sequence from the returned DNAfeature objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_id  feature_type  qualifier:label  start  end   strand  \n",
      "100         primer_bind   hU6-F            0      21    +       \n",
      "300         primer_bind   LKO.1 5'         171    191   +       \n",
      "1200        primer_bind   BGH-rev          5524   5542  -       \n",
      "1700        primer_bind   F1ori-R          6048   6068  -       \n",
      "1800        primer_bind   F1ori-F          6258   6280  +       \n",
      "1900        primer_bind   pRS-marker       6433   6453  -       \n",
      "2000        primer_bind   pGEX 3'          6552   6575  +       \n",
      "2100        primer_bind   pBRforEco        6612   6631  -       \n",
      "2400        primer_bind   Amp-R            7021   7041  -       \n",
      "2600        primer_bind   pBR322ori-F      8323   8343  +       \n",
      "\n",
      "feature_id  feature_type  qualifier:label  start  end   strand  \n",
      "1700        primer_bind   F1ori-R          6048   6068  -       \n",
      "2400        primer_bind   Amp-R            7021   7041  -       \n",
      "\n"
     ]
    }
   ],
   "source": [
    "feature_list = plasmid.searchfeature(key_attribute=\"feature_type\", query=\"primer_bind\")\n",
    "plasmid.printfeature(feature_list)\n",
    "sub_feature_list = plasmid.searchfeature(key_attribute=\"qualifier:label\", query=\".+-R$\", source=feature_list)\n",
    "plasmid.printfeature(sub_feature_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "##### Example code 11: Cut pX330 plasmid at multiple positions\n",
    "Cut a circular plasmid px330 at the three different positions, resulting in the generation of three fragments. Then, cut one of the three fragments again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_0', length='8484 bp', topology='circular'>\n",
      "[<queen.QUEEN object; project='pX330_1', length='1000 bp', topology='linear'>, <queen.QUEEN object; project='pX330_2', length='2000 bp', topology='linear'>, <queen.QUEEN object; project='pX330_3', length='5484 bp', topology='linear'>]\n",
      "<queen.QUEEN object; project='pX330_4', length='500 bp', topology='linear'>\n",
      "<queen.QUEEN object; project='pX330_5', length='1500 bp', topology='linear'>\n"
     ]
    }
   ],
   "source": [
    "print(plasmid)\n",
    "fragments = cutdna(plasmid ,1000, 2000, 4000)\n",
    "print(fragments)\n",
    "fragment3, fragment4 = cutdna(fragments[1], 500)\n",
    "print(fragment3)\n",
    "print(fragment4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----\n",
    "##### Example code 12: Digest pX330 plasmid by EcoRI  \n",
    "Digestion of pX330 plasmid with EcoRI can be simulated as follows.\n",
    "1. Search for EcoRI recognition sequence from the pX330 plasmid with RE site representation.\n",
    "2. Cut px330 sequence at the cut site(s)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_6', length='8488 bp', topology='linear'>\n",
      "5' AATTCCTAGA...AGTAAG---- 3'\n",
      "3' ----GGATCT...TCATTCTTAA 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sites     = plasmid.searchsequence(\"G^AATT_C\")\n",
    "fragments = cutdna(plasmid, *sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following codes generates the same product."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_7', length='8488 bp', topology='linear'>\n",
      "5' AATTCCTAGA...AGTAAG---- 3'\n",
      "3' ----GGATCT...TCATTCTTAA 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sites     = plasmid.searchsequence(\"GAATTC(-5/-1)\")\n",
    "fragments = cutdna(plasmid, *sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_8', length='8488 bp', topology='linear'>\n",
      "5' AATTCCTAGA...AGTAAG---- 3'\n",
      "3' ----GGATCT...TCATTCTTAA 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sites     = plasmid.searchsequence(\"(-1/-5)GAATTC\")\n",
    "fragments = cutdna(plasmid, *sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, QUEEN module has a collection of RE sites provided by New England Biolab. The collection can be used by importing REsites variable from “QUEEN/RE.py.” as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_9', length='8488 bp', topology='linear'>\n",
      "5' AATTCCTAGA...AGTAAG---- 3'\n",
      "3' ----GGATCT...TCATTCTTAA 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from QUEEN import cutsite #Import a collection of RE sites\n",
    "sites = plasmid.searchsequence(cutsite.lib[\"EcoRI\"])\n",
    "fragments = cutdna(plasmid, *sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----\n",
    "##### Example code 13: Digest pX330 plasmid by Type-IIS restriction enzyme BbsI \n",
    "Digestion of the pX330 plasmid by BbsI can also be simulated in the same way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_10', length='8466 bp', topology='linear'>\n",
      "5' GTTTTAGAGC...ACGAAA---- 3'\n",
      "3' ----ATCTCG...TGCTTTGTGG 5'\n",
      "\n",
      "<queen.QUEEN object; project='pX330_11', length='26 bp', sequence='CACCGGGTCTTCGAGAAGACCTGTTT', topology='linear'>\n",
      "5' CACCGGGTCT...AGACCT---- 3'\n",
      "3' ----CCCAGA...TCTGGACAAA 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sites = plasmid.searchsequence(\"GAAGAC(2/6)\")\n",
    "fragments = cutdna(plasmid,*sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following codes generates same products."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_12', length='26 bp', sequence='CACCGGGTCTTCGAGAAGACCTGTTT', topology='linear'>\n",
      "5' CACCGGGTCT...AGACCT---- 3'\n",
      "3' ----CCCAGA...TCTGGACAAA 5'\n",
      "\n",
      "<queen.QUEEN object; project='pX330_13', length='8466 bp', topology='linear'>\n",
      "5' GTTTTAGAGC...ACGAAA---- 3'\n",
      "3' ----ATCTCG...TGCTTTGTGG 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sites = plasmid.searchsequence(\"(6/2)GTCTTC\")\n",
    "fragments = cutdna(plasmid,*sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_14', length='8466 bp', topology='linear'>\n",
      "5' GTTTTAGAGC...ACGAAA---- 3'\n",
      "3' ----ATCTCG...TGCTTTGTGG 5'\n",
      "\n",
      "<queen.QUEEN object; project='pX330_15', length='26 bp', sequence='CACCGGGTCTTCGAGAAGACCTGTTT', topology='linear'>\n",
      "5' CACCGGGTCT...AGACCT---- 3'\n",
      "3' ----CCCAGA...TCTGGACAAA 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sites = plasmid.searchsequence(\"GAAGACNN^NNNN_\")\n",
    "fragments = cutdna(plasmid,*sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_16', length='26 bp', sequence='CACCGGGTCTTCGAGAAGACCTGTTT', topology='linear'>\n",
      "5' CACCGGGTCT...AGACCT---- 3'\n",
      "3' ----CCCAGA...TCTGGACAAA 5'\n",
      "\n",
      "<queen.QUEEN object; project='pX330_17', length='8466 bp', topology='linear'>\n",
      "5' GTTTTAGAGC...ACGAAA---- 3'\n",
      "3' ----ATCTCG...TGCTTTGTGG 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sites = plasmid.searchsequence(\"^NNNN_NNGTCTTC\")\n",
    "fragments = cutdna(plasmid,*sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, BbsI cut site also can be imported from \"Queen/RE.py\" as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_18', length='8466 bp', topology='linear'>\n",
      "5' GTTTTAGAGC...ACGAAA---- 3'\n",
      "3' ----ATCTCG...TGCTTTGTGG 5'\n",
      "\n",
      "<queen.QUEEN object; project='pX330_19', length='26 bp', sequence='CACCGGGTCTTCGAGAAGACCTGTTT', topology='linear'>\n",
      "5' CACCGGGTCT...AGACCT---- 3'\n",
      "3' ----CCCAGA...TCTGGACAAA 5'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sites = plasmid.searchsequence(cutsite.lib[\"BbsI\"])\n",
    "fragments = cutdna(plasmid,*sites)\n",
    "for fragment in fragments:\n",
    "    print(fragment)\n",
    "    fragment.printsequence(display=True, hide_middle=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "##### Example code 14: Crop a fragmented dna object in a specific region \n",
    "To retrieve only the 2nd fragment of ```fragments``` in \"Example code 11\", ```cropdna()``` is available instead of ```cutdna()```."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_20', length='2000 bp', topology='linear'>\n"
     ]
    }
   ],
   "source": [
    "fragment = cropdna(plasmid ,2000, 4000)\n",
    "print(fragment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "##### Example code 15: Trim single-stranded DNA on both ends to generate sticky ends  \n",
    "Sticky ends can be generated by trimming single-stranded nucleotide sequences when their end structures are given by top and bottom strand strings with \"\\*\" and \"-\" separated by \"/.\" The letters \"-\" show nucleotide letters being trimmed and “\\*” shod nucleotide letters being remained. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5' TACAAAATACGTGACGTAGA 3'\n",
      "3' ATGTTTTATGCACTGCATCT 5'\n",
      "\n",
      "5' -----AATACGTGACGTAGA 3'\n",
      "3' ATGTTTTATGCACTGCAT-- 5'\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'-----AATACGTGACGTAGA/ATGTTTTATGCACTGCAT--'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragment = cropdna(plasmid, 100, 120)\n",
    "fragment.printsequence(display=True)\n",
    "fragment = modifyends(fragment, \"-----/*****\", \"**/--\")\n",
    "fragment.printsequence(display=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code also can execute same process with the above one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5' -----AATACGTGACGTAGA 3'\n",
      "3' ATGTTTTATGCACTGCAT-- 5'\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'-----AATACGTGACGTAGA/ATGTTTTATGCACTGCAT--'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragment = cropdna(plasmid,'105/100', '120/118')\n",
    "fragment.printsequence(display=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A regex-like format can be used for the end stracture specification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5' -----AATACGTGACGTAGA 3'\n",
      "3' ATGTTTTATGCACTGCAT-- 5'\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'-----AATACGTGACGTAGA/ATGTTTTATGCACTGCAT--'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragment = modifyends(fragment, \"-{5}/*{5}\",\"*{2}/-{2}\")\n",
    "fragment.printsequence(display=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "##### Example code 16: Add additional sequences to both ends\n",
    "Addtional end sequence structures can also be given using \"modifyends\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5' TACATGCTACAAAATACGTGACGTAGATACGATG 3'\n",
      "3' ATGTACGATGTTTTATGCACTGCATCTATGCTAC 5'\n",
      "\n",
      "5' ---ATGCTACAAAATACGTGACGTAGATACG--- 3'\n",
      "3' ATGTACGATGTTTTATGCACTGCATCTATGCTAC 5'\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'---ATGCTACAAAATACGTGACGTAGATACG---/ATGTACGATGTTTTATGCACTGCATCTATGCTAC'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Add blunt-ended dsDNA sequences to both ends\n",
    "fragment = cropdna(plasmid, 100, 120)\n",
    "fragment = modifyends(fragment,\"TACATGC\",\"TACGATG\")\n",
    "fragment.printsequence(display=True)\n",
    "\n",
    "#Add sticky-ended dsDNA sequences to both ends\n",
    "fragment = cropdna(plasmid, 100, 120)\n",
    "fragment = modifyends(fragment,\"---ATGC/ATGTACG\",\"TACG---/ATGCTAC\")\n",
    "fragment.printsequence(display=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----- \n",
    "##### Example code 17: Insert EGFP sequence into pX330 using ```cutdna```, ```modifyends``` and, ```joindna```.\n",
    "1. Generate QUEEN object for the insert fragment EGFP by loading \"input/EGFP.fasta\".  \n",
    "2. Add EcoRI sites to both ends of EGFP fragment.\n",
    "3. Digest pX330 plasmid by EcoRI.  \n",
    "4. Join pX330 and EGFP fragment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5' AATTCGGCAG...ACAAGG---- 3'\n",
      "3' ----GCCGTC...TGTTCCTTAA 5'\n",
      "\n",
      "5' AATTCCTAGA...AGTAAG---- 3'\n",
      "3' ----GGATCT...TCATTCTTAA 5'\n",
      "\n",
      "<queen.QUEEN object; project='pX330_30', length='8488 bp', topology='linear'>\n",
      "<queen.QUEEN object; project='EGFP_2', length='787 bp', topology='linear'>\n",
      "<queen.QUEEN object; project='pX330_33', length='9267 bp', topology='circular'>\n"
     ]
    }
   ],
   "source": [
    "EGFP     = QUEEN(record=\"input/EGFP.fasta\")\n",
    "#Add EcoRI site to both ends\n",
    "EGFP     = modifyends(EGFP, cutsite.lib[\"EcoRI\"].seq, cutsite.lib[\"EcoRI\"].seq)\n",
    "#Digest EGFP\n",
    "sites    = EGFP.searchsequence(cutsite.lib[\"EcoRI\"]) \n",
    "insert   = cutdna(EGFP, *sites)[1]\n",
    "insert.printsequence(display=True, hide_middle=10)\n",
    "#Digest pX330 \n",
    "sites    = plasmid.searchsequence(cutsite.lib[\"EcoRI\"])\n",
    "backbone = cutdna(plasmid, *sites)[0]\n",
    "backbone.printsequence(display=True, hide_middle=10)\n",
    "#Ligation\n",
    "pEGFP    = joindna(backbone, insert, topology=\"circular\") \n",
    "print(backbone)\n",
    "print(insert)\n",
    "print(pEGFP)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "##### Example code 18: Insert gRNA sequence into pX330\n",
    "The pX330 plasmid gives gRNA insert site flanked by two inverted BbsI cut sites. By the digestion at BbsI sites, any gRNA sequence can be cloned into the pX330 as follows. For details, see https://www.addgene.org/crispr/zhang/\n",
    "1. Generate QUEEN object from a gRNA sequence input.  \n",
    "2. Add proper sticky-end sequences that can be joined with the pX330 plasmid digested by BbsI to both ends of the gRNA sequence.\n",
    "3. Digest pX330 plasmid by BbsI. \n",
    "4. Join pX300 and the gRNA sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_34', length='8466 bp', topology='linear'>\n",
      "<queen.QUEEN object; project='gRNA_0', length='29 bp', sequence='CACCGACCATTGTTCAATATCGTCCGTTT', topology='linear'>\n",
      "<queen.QUEEN object; project='pX330_38', length='8487 bp', topology='circular'>\n"
     ]
    }
   ],
   "source": [
    "gRNA      = QUEEN(seq=\"ACCATTGTTCAATATCGTCC\", project=\"gRNA\") \n",
    "insert    = modifyends(gRNA, \"CACCG/----C\", \"----/CAAA\")\n",
    "sites     = plasmid.searchsequence(cutsite.lib[\"BbsI\"])\n",
    "fragments = cutdna(plasmid, *sites)\n",
    "backbone  = fragments[0] if len(fragments[0].seq) > len(fragments[1].seq) else fragment[1]\n",
    "pgRNA     = joindna(backbone, insert, topology=\"circular\")\n",
    "print(backbone)\n",
    "print(insert)\n",
    "print(pgRNA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "##### Example code 19: Flip ampicillin resistant gene in px330 plasmid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_id  feature_type  qualifier:label  start  end   strand  \n",
      "2300        CDS           AmpR             6803   7664  +       \n",
      "\n",
      "feature_id  feature_type  qualifier:label  start  end   strand  \n",
      "2400        CDS           AmpR             6803   7664  -       \n",
      "\n",
      "################################################################################\n",
      "#This source code was auto-generated by 'quine' funtion of QUEEN 1.0.0.\n",
      "#Project Name    :dna_59\n",
      "#File Name       :None\n",
      "#Creation Date   :2021-09-14\n",
      "################################################################################\n",
      "project='dna_59'\n",
      "import sys\n",
      "sys.path.append(\"/Users/hideto/Dropbox (Yachie Lab)/HIDETO_MORI.LAB/Experiments/Project/Dbrick/github/demo/tutorial/../..\")\n",
      "from QUEEN.queen import *\n",
      "from QUEEN import cutsite as cs\n",
      "\n",
      "QUEEN.dna_dict['pX330_0'] = QUEEN(record='https://benchling.com/s/seq-U4pePb09KHutQzjyOPQV', dbtype='benchling', process_id='dna_59-ZudkSOB2S7', originals=[])\n",
      "QUEEN.queried_features_dict['pX330_17'] = QUEEN.dna_dict['pX330_0'].searchfeature(key_attribute='all', query='^AmpR$', process_id='dna_59-7lVKtWXy3H', originals=[])\n",
      "QUEEN.dna_dict['pX330_39'], QUEEN.dna_dict['pX330_40'] = cutdna(QUEEN.dna_dict['pX330_0'], QUEEN.queried_features_dict['pX330_17'][0].start, QUEEN.queried_features_dict['pX330_17'][0].end, process_id='dna_59-zmRnczVwK0', originals=[])\n",
      "QUEEN.dna_dict['dna_3'] = flipdna(QUEEN.dna_dict['pX330_39'], process_id='dna_59-DTs3zSqpYE', originals=[])\n",
      "QUEEN.dna_dict['dna_59'] = joindna(*[QUEEN.dna_dict['dna_3'], QUEEN.dna_dict['pX330_40']], topology='circular', process_id='dna_59-1u2Im5u23F', originals=[])\n",
      "if __name__ == '__main__':\n",
      "    check = quine(QUEEN.dna_dict['dna_59'], author=None, project=project, _check=True)\n",
      "    if check == True:\n",
      "        QUEEN.dna_dict['dna_59'].writedna()\n"
     ]
    }
   ],
   "source": [
    "site         = plasmid.searchfeature(query=\"^AmpR$\")[0]\n",
    "fragments    = cutdna(plasmid, site.start, site.end)\n",
    "fragments[0] = flipdna(fragments[0])\n",
    "new_plasmid  = joindna(*fragments, topology=\"circular\")\n",
    "plasmid.printfeature(plasmid.searchfeature(query=\"^AmpR$\"))\n",
    "new_plasmid.printfeature(new_plasmid.searchfeature(query=\"^AmpR$\")) \n",
    "quine(new_plasmid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Example code 20: Insert EGFP sequence into pX330 using \"editsequence().\"**  \n",
    "The molecular cloning process described in \"Example code 17\" can be reproduced with more simple codes uding ```editsequence```. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<queen.QUEEN object; project='pX330_0', length='8484 bp', topology='circular'>\n",
      "<queen.QUEEN object; project='pX330_45', length='9267 bp', topology='linear'>\n"
     ]
    }
   ],
   "source": [
    "EGFP  = QUEEN(record=\"input/EGFP.fasta\")\n",
    "pEGFP = editsequence(plasmid, \"({})\".format(cutsite.lib[\"EcoRI\"].seq), r\"\\1{}\\1\".format(EGFP.seq))\n",
    "print(plasmid)\n",
    "print(pEGFP)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----\n",
    "##### Example code 21: Insert \"AAAAA\" into the head of CDS features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1231\t1302\tGACTAAAAAATAAGGACCAC\t3xFLAG\n",
      "1308\t1334\tCCAAAAAAAAGAAGAAGCGG\tSV40 NLS\n",
      "1358\t5464\tGACAAAAAAAGAAGTACAGC\tCas9\n",
      "5464\t5517\tAAAAGAAAAAGCCGGCGGCC\tnucleoplasmin NLS\n",
      "6823\t7689\tATGAGAAAAATATTCAACAT\tAmpR\n"
     ]
    }
   ],
   "source": [
    "new_plasmid  = editfeature(plasmid, key_attribute=\"feature_type\", query=\"CDS\", strand=1, target_attribute=\"sequence\", operation=replaceattribute(r\"([ATGC]{5})(.+)\", r\"\\1AAAAA\\2\"))\n",
    "for feat in new_plasmid.searchfeature(key_attribute=\"feature_type\", query=\"CDS\", strand=1):\n",
    "    print(feat.start, feat.end, new_plasmid.printsequence(feat.start, feat.start+20, strand=1), feat.qualifiers[\"label\"][0], sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pX330\n",
      "1231\t1302\tAAAAAGACTATAAGGACCAC\t3xFLAG\n",
      "1308\t1334\tAAAAACCAAAGAAGAAGCGG\tSV40 NLS\n",
      "1358\t5464\tAAAAAGACAAGAAGTACAGC\tCas9\n",
      "5464\t5517\tAAAAAAAAAGGCCGGCGGCC\tnucleoplasmin NLS\n",
      "6823\t7689\tAAAAAATGAGTATTCAACAT\tAmpR\n",
      "feature_id  feature_type   qualifier:label     start  end   strand  \n",
      "1           primer_bind    hU6-F               0      21    +       \n",
      "100         promoter       U6 promoter         0      241   +       \n",
      "200         source         source              0      1231  +       \n",
      "300         primer_bind    LKO.1 5'            171    191   +       \n",
      "400         misc_RNA       gRNA scaffold       267    343   +       \n",
      "500         enhancer       CMV enhancer        439    725   +       \n",
      "600         intron         hybrid intron       983    1211  +       \n",
      "700         regulatory     Kozak sequence      1222   1231  +       \n",
      "800         CDS            3xFLAG              1231   1302  +       \n",
      "900         regulatory     Kozak sequence      1236   1237  +       \n",
      "1000        source         source              1236   1308  +       \n",
      "1100        CDS            SV40 NLS            1308   1334  +       \n",
      "1200        source         source              1313   1358  +       \n",
      "1300        CDS            Cas9                1358   5464  +       \n",
      "1400        source         source              1363   5464  +       \n",
      "1500        CDS            nucleoplasmin NLS   5464   5517  +       \n",
      "1600        source         source              5469   6823  +       \n",
      "1700        primer_bind    BGH-rev             5544   5562  -       \n",
      "1800        polyA_signal   bGH poly(A) signal  5550   5758  +       \n",
      "1900        repeat_region  AAV2 ITR            5766   5896  +       \n",
      "2000        repeat_region  AAV2 ITR            5766   5907  +       \n",
      "2100        rep_origin     f1 ori              5981   6437  +       \n",
      "2200        primer_bind    F1ori-R             6068   6088  -       \n",
      "2300        primer_bind    F1ori-F             6278   6300  +       \n",
      "2400        primer_bind    pRS-marker          6453   6473  -       \n",
      "2500        primer_bind    pGEX 3'             6572   6595  +       \n",
      "2600        primer_bind    pBRforEco           6632   6651  -       \n",
      "2700        promoter       AmpR promoter       6718   6823  +       \n",
      "2900        CDS            AmpR                6823   7689  +       \n",
      "2800        source         source              6828   8509  +       \n",
      "3000        primer_bind    Amp-R               7046   7066  -       \n",
      "3100        rep_origin     ori                 7859   8448  +       \n",
      "3200        primer_bind    pBR322ori-F         8348   8368  +       \n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(plasmid._unique_id)\n",
    "new_plasmid  = editfeature(plasmid, key_attribute=\"feature_type\", query=\"CDS\", strand=1, target_attribute=\"sequence\", operation=replaceattribute(r\"(.+)\", r\"AAAAA\\1\"))\n",
    "for feat in new_plasmid.searchfeature(key_attribute=\"feature_type\", query=\"CDS\", strand=1):\n",
    "    print(feat.start, feat.end, new_plasmid.printsequence(feat.start, feat.start+20, strand=1), feat.qualifiers[\"label\"][0], sep=\"\\t\")\n",
    "new_plasmid.printfeature()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----\n",
    "##### Example code 22: Convert feature type from \"CDS\" to \"gene\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_id  feature_type   qualifier:label     start  end   strand  \n",
      "1           source         source              0      8484  +       \n",
      "100         primer_bind    hU6-F               0      21    +       \n",
      "200         promoter       U6 promoter         0      241   +       \n",
      "300         primer_bind    LKO.1 5'            171    191   +       \n",
      "400         misc_RNA       gRNA scaffold       267    343   +       \n",
      "500         enhancer       CMV enhancer        439    725   +       \n",
      "600         intron         hybrid intron       983    1211  +       \n",
      "700         regulatory     Kozak sequence      1222   1232  +       \n",
      "800         gene           3xFLAG              1231   1297  +       \n",
      "900         gene           SV40 NLS            1303   1324  +       \n",
      "1000        gene           Cas9                1348   5449  +       \n",
      "1100        gene           nucleoplasmin NLS   5449   5497  +       \n",
      "1200        primer_bind    BGH-rev             5524   5542  -       \n",
      "1300        polyA_signal   bGH poly(A) signal  5530   5738  +       \n",
      "1400        repeat_region  AAV2 ITR            5746   5876  +       \n",
      "1500        repeat_region  AAV2 ITR            5746   5887  +       \n",
      "1600        rep_origin     f1 ori              5961   6417  +       \n",
      "1700        primer_bind    F1ori-R             6048   6068  -       \n",
      "1800        primer_bind    F1ori-F             6258   6280  +       \n",
      "1900        primer_bind    pRS-marker          6433   6453  -       \n",
      "2000        primer_bind    pGEX 3'             6552   6575  +       \n",
      "2100        primer_bind    pBRforEco           6612   6631  -       \n",
      "2200        promoter       AmpR promoter       6698   6803  +       \n",
      "2300        gene           AmpR                6803   7664  +       \n",
      "2400        primer_bind    Amp-R               7021   7041  -       \n",
      "2500        rep_origin     ori                 7834   8423  +       \n",
      "2600        primer_bind    pBR322ori-F         8323   8343  +       \n",
      "\n"
     ]
    }
   ],
   "source": [
    "new_plasmid = editfeature(plasmid, key_attribute=\"feature_type\", query=\"CDS\", target_attribute=\"feature_type\", operation=replaceattribute(\"gene\"))\n",
    "new_plasmid.printfeature()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "##### Example code 23: Annotate recognition sites for unique cutters on pX330\n",
    "1. Search for the unique cutters in pX330 from the restriction enzymes provided by NEW England Biolabs.\n",
    "2. Add sequence features for the unique cutters into pX330.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_id  feature_type  qualifier:label  start  end   strand  sequence      \n",
      "RE-1        misc_bind     Acc65I           433    439   +       GGTACC        \n",
      "RE-2        misc_bind     AgeI             1216   1222  +       ACCGGT        \n",
      "RE-3        misc_bind     ApaI             2700   2706  +       GGGCCC        \n",
      "RE-4        misc_bind     BglII            1595   1601  +       AGATCT        \n",
      "RE-5        misc_bind     BsaBI            4839   4849  +       GATCACCATC    \n",
      "RE-6        misc_bind     BseRI            1098   1104  -       GAGGAG        \n",
      "RE-7        misc_bind     BsmI             4979   4985  +       GAATGC        \n",
      "RE-8        misc_bind     CspCI            4127   4139  +       CAAAGCACGTGG  \n",
      "RE-9        misc_bind     EcoRI            5500   5506  +       GAATTC        \n",
      "RE-10       misc_bind     EcoRV            3196   3202  +       GATATC        \n",
      "RE-11       misc_bind     FseI             5472   5480  +       GGCCGGCC      \n",
      "RE-12       misc_bind     FspI             7365   7371  +       TGCGCA        \n",
      "RE-13       misc_bind     KasI             5887   5893  +       GGCGCC        \n",
      "RE-16       misc_bind     NotI             5738   5746  +       GCGGCCGC      \n",
      "RE-17       misc_bind     PaqCI            1184   1191  +       CACCTGC       \n",
      "RE-19       misc_bind     PmlI             4132   4138  +       CACGTG        \n",
      "RE-20       misc_bind     PsiI             6317   6323  +       TTATAA        \n",
      "RE-22       misc_bind     PvuI             7218   7224  +       CGATCG        \n",
      "RE-23       misc_bind     SacII            7522   7528  +       CCGCGG        \n",
      "RE-24       misc_bind     SbfI             5879   5887  +       CCTGCAGG      \n",
      "RE-26       misc_bind     SnaBI            698    704   +       TACGTA        \n",
      "RE-27       misc_bind     XbaI             427    433   +       TCTAGA        \n",
      "\n"
     ]
    }
   ],
   "source": [
    "unique_cutters = []\n",
    "for key, re in cutsite.lib.items():\n",
    "    sites = plasmid.searchsequence(re.cutsite)\n",
    "    if len(sites) == 1: \n",
    "        unique_cutters.append(sites[0])\n",
    "    else:\n",
    "        pass \n",
    "new_plasmid = editfeature(plasmid, source=unique_cutters, target_attribute=\"feature_id\", operation=createattribute(\"RE\"))\n",
    "new_plasmid = editfeature(new_plasmid, key_attribute=\"feature_id\", query=\"RE\", target_attribute=\"feature_type\", operation=replaceattribute(\"misc_bind\"))\n",
    "features    = new_plasmid.searchfeature(key_attribute=\"feature_type\", query=\"misc_bind\")\n",
    "new_plasmid.printfeature(features, seq=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Example code 19’: Modified “Example code 19” "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_id  feature_type  qualifier:label  start  end   strand  \n",
      "2300        CDS           AmpR             6803   7664  +       \n",
      "\n",
      "feature_id  feature_type  qualifier:label  start  end   strand  \n",
      "2400        CDS           AmpR             6803   7664  -       \n",
      "\n",
      "################################################################################\n",
      "#This source code was auto-generated by 'quine' funtion of QUEEN 1.0.0.\n",
      "#Project Name    :new_plasmid\n",
      "#File Name       :None\n",
      "#Creation Date   :2021-09-13\n",
      "################################################################################\n",
      "project='new_plasmid'\n",
      "import sys\n",
      "sys.path.append(\"/Users/hideto/dropbox/HIDETO_MORI.LAB/Experiments/Project/Dbrick/github\")\n",
      "from QUEEN.queen import *\n",
      "from QUEEN import cutsite as cs\n",
      "set_namespace(globals())\n",
      "\n",
      "QUEEN(record='input/pX330.gbk', project='plasmid', product='plasmid', process_id='new_plasmid-85osVkUdIC', originals=[])\n",
      "plasmid.searchfeature(key_attribute='all', query='^AmpR$', product='sites', process_id='new_plasmid-mbRHHmVX0d', originals=[])\n",
      "cutdna(plasmid, sites[0].start, sites[0].end, product='fragments', process_id='new_plasmid-6UDM2HrKQK', originals=[])\n",
      "flipdna(fragments[0], product='fragments0_rc', process_id='new_plasmid-52ONFMhYTf', originals=[])\n",
      "joindna(*[fragments0_rc, fragments[1]], topology='circular', product='new_plasmid', process_id='new_plasmid-eVLxRtdESR', originals=[])\n",
      "if __name__ == '__main__':\n",
      "    check = quine(new_plasmid, author=None, project=project, _check=True)\n",
      "    if check == True:\n",
      "        new_plasmid.writedna()\n"
     ]
    }
   ],
   "source": [
    "import sys \n",
    "sys.path.append(\"../../\")\n",
    "from QUEEN.queen import *\n",
    "set_namespace(globals()) \n",
    "QUEEN(record=\"input/pX330.gbk\", product=\"plasmid\")\n",
    "plasmid.searchfeature(query=\"^AmpR$\", product=\"sites\")\n",
    "cutdna(plasmid, sites[0].start, sites[0].end, product=\"fragments\")\n",
    "flipdna(fragments[0], product=\"fragments0_rc\")\n",
    "joindna(fragments0_rc, fragments[1], topology=\"circular\", product=\"new_plasmid\")\n",
    "plasmid.printfeature(plasmid.searchfeature(query=\"^AmpR$\"))\n",
    "new_plasmid.printfeature(new_plasmid.searchfeature(query=\"^AmpR$\")) \n",
    "quine(new_plasmid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
